### DISCUSSION TITLE
Use typescript to build an Apollo Server

### DISCUSSION THREAD

**Discussion by rntjr** [apollographql/apollo-server#4559] [Q&A]
I am starting in GraphQL with Apollo.
I want to develop Apollo Server with typescript and needed some example. I also needed an example of how to put schemas and resolvers together and make the server modular.

I checked the documentation and it is possible to use graphql-tools, natively also the apollo federation. But I still needed an example to be able to create a design pattern for the project.

**Comment by KeithGillette** [comment]
Have you looked at [NestJS + GraphQL](https://docs.nestjs.com/graphql/quick-start) or [TypeGraphQL](https://typegraphql.com/)? They have clean, modular patterns.

We have a TypeScript-based Apollo Server implementation with a custom architecture that has fairly good separation of concerns, but is definitely not as modular as either of those libraries. Our high-level class structure is:

- GQLServerService: configures & instantiates Apollo Server, injects GQLServerService & SchemaService
- SchemaService: Loads & returns the `graphqls` schema for GQLServerService
- ResolverMapService: Returns the resolver map to GQLServerService, injects EntityResolversServices
- EntityResolversService: 1 class per domain entity with methods providing the Resolver functions that inject and call EntityReadQueryServices (query handlers) and EntityActionMutationServices (command handlers)
- EntityReadQueryService: 1 class per Query with `handle` method that validates (via `class-validator` decorator), authorizes, & executes the Query constructed by the calling EntityResolversService method
- EntityActionMutationService: 1 class per Mutation with `handle` method that validates (via `class-validator` decorator), authorizes, & executes the Mutation constructed by the calling EntityResolversService method
- EntityPropertyUpdatedService: 1 class per Subscription with `handle` method that validates (via `class-validator` decorator) & returns an AsyncIterator for the Subscription constructed by the calling EntityResolversService method
